### PR TITLE
Port FxCop rule CA1820: Test for empty strings using string length (http...

### PR DESCRIPTION
**Pull request by mavasani**
...s://msdn.microsoft.com/library/ms182279.aspx) to System.Runtime analyzer.

### CODE REVIEW

**Comment by mavasani** [comment]
@srivatsn @tmeschter @ManishJayaswal @shyamnamboodiripad @heejaechang @jmarolf  please review.

I am going to port another FXCop rule and then work on fixers for both of them together.

File: src/Diagnostics/FxCop/System.Runtime.Analyzers/Core/Performance/TestForEmptyStringsUsingStringLength.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+
+            var symbol = model.GetSymbolInfo(expression, cancellationToken).Symbol as IFieldSymbol;
+            if (symbol != null &&
```
Comment (pharring) [inline src/Diagnostics/FxCop/System.Runtime.Analyzers/Core/Performance/TestForEmptyStringsUsingStringLength.cs]: Turn ```if (expr) return true; else return false;``` into a "return" i.e.

```C#
return symbol != null &&
       symbol.Name.Equals(StringEmptyFieldName) && etc...

File: src/Diagnostics/FxCop/System.Runtime.Analyzers/Core/Performance/TestForEmptyStringsUsingStringLength.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        {
+            if (expression == null)
+            {
+                return false;
+            }
+
+            var constantValueOpt = model.GetConstantValue(expression, cancellationToken);
+            if (constantValueOpt.HasValue)
+            {
+                if (constantValueOpt.Value != null)
```
Comment (pharring) [inline src/Diagnostics/FxCop/System.Runtime.Analyzers/Core/Performance/TestForEmptyStringsUsingStringLength.cs]: Does this work?
```C#
return (constantValueOpt.Value as string)?.Length == 0;

File: src/Diagnostics/FxCop/System.Runtime.Analyzers/Core/Performance/TestForEmptyStringsUsingStringLength.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+                                                                             DiagnosticCategory.Performance,
+                                                                             DiagnosticSeverity.Warning,
+                                                                             isEnabledByDefault: false,
+                                                                             description: s_localizableDescription,
+                                                                             helpLinkUri: "https://msdn.microsoft.com/library/ms182279.aspx",
+                                                                             customTags: WellKnownDiagnosticTags.Telemetry);
+
+        protected const string StringEmptyFieldName = "Empty";
+
+        protected sealed override DiagnosticDescriptor Descriptor
```
Comment (pharring) [inline src/Diagnostics/FxCop/System.Runtime.Analyzers/Core/Performance/TestForEmptyStringsUsingStringLength.cs]: Save some typing and use an expression bodied property:
```C#
   protected sealed override DiagnosticDescriptor Descriptor => Rule;

File: src/Diagnostics/FxCop/System.Runtime.Analyzers/Core/AbstractSyntaxNodeAnalyzer.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace System.Runtime.Analyzers
+{
+    public abstract class AbstractSyntaxNodeAnalyzer<TLanguageKindEnum> : DiagnosticAnalyzer where TLanguageKindEnum : struct
+    {
+        public sealed override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
```
Comment (pharring) [inline src/Diagnostics/FxCop/System.Runtime.Analyzers/Core/AbstractSyntaxNodeAnalyzer.cs]:
```C#
public sealed override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Descriptor);

File: src/Diagnostics/FxCop/System.Runtime.Analyzers/CSharp/Performance/CSharpTestForEmptyStringsUsingStringLengthAnalyzer.cs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+                    break;
+            }
+        }
+
+        private void AnalyzeInvocationExpression(SyntaxNodeAnalysisContext context)
+        {
+            var node = (InvocationExpressionSyntax)context.Node;
+            if (node.Expression.IsKind(SyntaxKind.SimpleMemberAccessExpression) &&
+                node.ArgumentList != null &&
+                node.ArgumentList.Arguments.Count > 0)
```
Comment (pharring) [inline src/Diagnostics/FxCop/System.Runtime.Analyzers/CSharp/Performance/CSharpTestForEmptyStringsUsingStringLengthAnalyzer.cs]:
```C#
node.ArgumentList?.Arguments.Count > 0

**Comment by tmeschter** [comment]
LGTM

**Comment by mavasani** [comment]
@pharring I have addressed your feedback.

**Comment by pharring** [comment]
Thanks. :+1:

**Pull request merged**